### PR TITLE
ROC-2559 debts page

### PR DESCRIPTION
**Pull request by yarpo**
### JIRA link
https://tools.hmcts.net/jira/browse/ROC-2559

### Change description
Added debts page

### Work checklist

- [x] Unit tests added where applicable
- [x] Route tests added for new pages
- [x] New pages included in a11y tests
- [x] Task list and task completeness checks updated
- [x] Check and send page updated
- [x] Routes, page content and page flows of new features are toggled 
- [x] UI changes look good on mobile
- [x] Required Google Analytics events are being sent 

### Developer self-QA run statement

- [x] I have clicked through the running application to see if all changes I made actually work.

### Does this PR introduce a breaking change?

- [x] Yes
- [ ] No

### CODE REVIEW

File: src/main/app/forms/validation/validators/minNumberOfPopulatedRows.ts [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+  ValidationOptions,
+  ValidatorConstraint,
+  ValidatorConstraintInterface
+} from 'class-validator'
+
+import { MultiRowForm } from 'forms/models/multiRowForm'
+import { MultiRowFormItem } from 'forms/models/multiRowFormItem'
+
+@ValidatorConstraint()
+export class MinNumberOfPopulatedRowsConstraint implements ValidatorConstraintInterface {
```
Comment (timja-kainos) [inline src/main/app/forms/validation/validators/minNumberOfPopulatedRows.ts]: `AtLeastOnePopulatedRow`?
The name here looks like it is configurable, but it isn't

File: src/main/features/response/form/models/statement-of-means/debtRow.ts [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+import { ValidationErrors as GlobalValidationErrors } from 'forms/validation/validationErrors'
+import { MultiRowFormItem } from 'forms/models/multiRowFormItem'
+import { Fractions } from 'forms/validation/validators/fractions'
+import { toNumberOrUndefined } from 'common/utils/numericUtils'
+import { IsNotBlank } from 'forms/validation/validators/isBlank'
+import { MaxLength } from 'forms/validation/validators/maxLengthValidator'
+import { ValidationConstraints } from 'forms/validation/validationConstraints'
+
+export class ValidationErrors {
+  static readonly DEBT_REQUIRED: string = 'Enter a value'
```
Comment (timja-kainos) [inline src/main/features/response/form/models/statement-of-means/debtRow.ts]: this doesn't look very specific to debts, maybe move it?

File: src/main/app/forms/validation/validators/minNumberOfPopulatedRows.ts (hunk 1 above)
Comment (yarpo) [reply src/main/app/forms/validation/validators/minNumberOfPopulatedRows.ts]: yeah, you're right

File: src/main/features/response/form/models/statement-of-means/debtRow.ts [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+  @IsDefined({ message: ValidationErrors.DEBT_REQUIRED })
+  @IsNotBlank({ message: ValidationErrors.DEBT_REQUIRED })
+  @MaxLength(ValidationConstraints.STANDARD_TEXT_INPUT_MAX_LENGTH, { message: GlobalValidationErrors.TEXT_TOO_LONG })
+  debt?: string
+
+  @ValidateIf(o => o.isAtLeastOneFieldPopulated())
+  @IsDefined({ message: GlobalValidationErrors.NUMBER_REQUIRED })
+  @IsDefined({ message: GlobalValidationErrors.NUMBER_REQUIRED })
+  @Fractions(0, 2, { message: GlobalValidationErrors.AMOUNT_INVALID_DECIMALS })
+  @Min(0, { message: GlobalValidationErrors.POSITIVE_NUMBER_REQUIRED })
```
Comment (timja-kainos) [inline src/main/features/response/form/models/statement-of-means/debtRow.ts]: we have `DefaultValidationErrors` and `GlobalValidationErrors` now :(
should just rename the class imo, not required for this change..

File: src/main/features/response/routes/statement-of-means/debts.ts [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+import { FeatureToggleGuard } from 'guards/featureToggleGuard'
+import { Debts } from 'response/form/models/statement-of-means/debts'
+
+const page: RoutablePath = StatementOfMeansPaths.debtsPage
+
+function renderView (form: Form<Debts>, res: express.Response): void {
+  res.render(page.associatedView, { form: form })
+}
+
+function actionHandler (req: express.Request, res: express.Response, next: express.NextFunction): void {
```
Comment (timja-kainos) [inline src/main/features/response/routes/statement-of-means/debts.ts]: looks like a generic actionHandler that could be extracted? up to you

File: src/main/features/response/views/statement-of-means/debts.njk [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+            radioOption(form = form, label = 'Yes', name = 'hasAnyDebts', value = 'true', checked = form.valueFor('hasAnyDebts') === true)
+          }}
+
+          {% set noRowsError = form.errorFor('rows') %}
+          <div class="form-group panel panel-border-narrow js-hidden {% if noRowsError %} form-group-error {% endif %}"
+               id="hasAnyDebts-true" aria-hidden="false">
+
+            {% if noRowsError %}<span class="error-message" id="rows[label]">{{ t(noRowsError) }}</span>{% endif %}
+
+            <div class="timeline-rows">
```
Comment (timja-kainos) [inline src/main/features/response/views/statement-of-means/debts.njk]: timeline-rows?

File: src/main/features/response/views/statement-of-means/debts.njk [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+                  <label class="form-label-bold">{{ t('Monthly payments') }}</label>
+                </div>
+              </div>
+
+              {% for row in form.model.rows %}
+                {{ debtRow(loop.index0, form) }}
+              {% endfor %}
+            </div>
+
+            {% if form.model and form.model.canAddMoreRows() %}
```
Comment (timja-kainos) [inline src/main/features/response/views/statement-of-means/debts.njk]: when do you not have a form model here?

File: src/test/app/forms/validation/validators/minNumberOfPopoulatedRows.ts [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+const POPULATED: boolean = false
+
+describe('MinNumberOfPopulatedRowsConstraint', () => {
+  const constraint: MinNumberOfPopulatedRowsConstraint = new MinNumberOfPopulatedRowsConstraint()
+
+  describe('validate', () => {
+
+    context('should accept', () => {
+
+      it('should accept null value', () => {
```
Comment (timja-kainos) [inline src/test/app/forms/validation/validators/minNumberOfPopoulatedRows.ts]: please don't test against null

File: src/main/features/response/views/statement-of-means/debts.njk (hunk 6 above)
Comment (yarpo) [reply src/main/features/response/views/statement-of-means/debts.njk]: `Form<Model> model = Form.empty()`

->
```
2017-12-04T16:47:14+00:00 ERROR errorLogger.js: Template render error: (/Users/patrykj/cmc/cmc-citizen-frontend/src/main/features/response/views/statement-of-means/debts.njk) [Line 98, Column 41]
  Error: Unable to call `form["model"]["canAddMoreRows"]`, which is undefined or falsey
```

File: src/test/app/forms/validation/validators/minNumberOfPopoulatedRows.ts (hunk 7 above)
Comment (yarpo) [reply src/test/app/forms/validation/validators/minNumberOfPopoulatedRows.ts]: You mean `undefined` or removing this test at all?

File: src/main/features/response/form/models/statement-of-means/debtRow.ts [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+  totalOwed?: number
+
+  @ValidateIf(o => o.isAtLeastOneFieldPopulated())
+  @IsDefined({ message: GlobalValidationErrors.NUMBER_REQUIRED })
+  @IsDefined({ message: GlobalValidationErrors.NUMBER_REQUIRED })
+  @Fractions(0, 2, { message: GlobalValidationErrors.AMOUNT_INVALID_DECIMALS })
+  @Min(0, { message: GlobalValidationErrors.POSITIVE_NUMBER_REQUIRED })
```
Comment (rlewan) [inline src/main/features/response/form/models/statement-of-means/debtRow.ts]: 0 is not a positive number :)

Comment (yarpo) [reply src/main/features/response/form/models/statement-of-means/debtRow.ts]: ok, `NON_NEGATIVE_NUMBER_REQUIRED`

File: src/main/features/response/form/models/statement-of-means/debtRow.ts [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+  totalOwed?: number
+
+  @ValidateIf(o => o.isAtLeastOneFieldPopulated())
+  @IsDefined({ message: GlobalValidationErrors.NUMBER_REQUIRED })
+  @IsDefined({ message: GlobalValidationErrors.NUMBER_REQUIRED })
```
Comment (rlewan) [inline src/main/features/response/form/models/statement-of-means/debtRow.ts]: Duplicated constraint.

File: src/main/features/response/form/models/statement-of-means/debtRow.ts [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+
+  @ValidateIf(o => o.isAtLeastOneFieldPopulated())
+  @IsDefined({ message: ValidationErrors.DEBT_REQUIRED })
+  @IsNotBlank({ message: ValidationErrors.DEBT_REQUIRED })
+  @MaxLength(ValidationConstraints.STANDARD_TEXT_INPUT_MAX_LENGTH, { message: GlobalValidationErrors.TEXT_TOO_LONG })
+  debt?: string
+
+  @ValidateIf(o => o.isAtLeastOneFieldPopulated())
+  @IsDefined({ message: GlobalValidationErrors.NUMBER_REQUIRED })
+  @IsDefined({ message: GlobalValidationErrors.NUMBER_REQUIRED })
```
Comment (rlewan) [inline src/main/features/response/form/models/statement-of-means/debtRow.ts]: Duplicated constraint.

File: src/main/features/response/form/models/statement-of-means/debts.ts [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+  static readonly ENTER_AT_LEAST_ONE_ROW: string = 'Enter at least one debt'
+}
+
+export class Debts extends MultiRowForm<DebtRow> {
+
+  @IsDefined({ message: GlobalValidationErrors.YES_NO_REQUIRED })
+  hasAnyDebts: boolean
+
+  @ValidateIf(o => o.hasAnyDebts === true)
+  @MinNumberOfPopulatedRows({ message: ValidationErrors.ENTER_AT_LEAST_ONE_ROW })
```
Comment (rlewan) [inline src/main/features/response/form/models/statement-of-means/debts.ts]: I would rename this to `@AtLeastOnePopulatedRow` or add a parameter.

File: src/test/app/forms/validation/validators/minNumberOfPopoulatedRows.ts (hunk 7 above)
Comment (timja-kainos) [reply src/test/app/forms/validation/validators/minNumberOfPopoulatedRows.ts]: use `undefined` if the test is valid

File: src/main/features/response/form/models/statement-of-means/debtRow.ts (hunk 8 above)
Comment (yarpo) [reply src/main/features/response/form/models/statement-of-means/debtRow.ts]: done

File: src/main/features/response/form/models/statement-of-means/debtRow.ts (hunk 9 above)
Comment (yarpo) [reply src/main/features/response/form/models/statement-of-means/debtRow.ts]: done

File: src/main/features/response/form/models/statement-of-means/debtRow.ts (hunk 10 above)
Comment (yarpo) [reply src/main/features/response/form/models/statement-of-means/debtRow.ts]: done

File: src/main/features/response/routes/statement-of-means/debts.ts (hunk 4 above)
Comment (yarpo) [reply src/main/features/response/routes/statement-of-means/debts.ts]: I will

File: src/main/features/response/form/models/statement-of-means/debtRow.ts (hunk 2 above)
Comment (yarpo) [reply src/main/features/response/form/models/statement-of-means/debtRow.ts]: I have talked to Tarik. All msg here will be updated

File: src/main/app/forms/validation/validators/minNumberOfPopulatedRows.ts (hunk 1 above)
Comment (yarpo) [reply src/main/app/forms/validation/validators/minNumberOfPopulatedRows.ts]: done

File: src/main/features/response/routes/statement-of-means/debts.ts (hunk 4 above)
Comment (yarpo) [reply src/main/features/response/routes/statement-of-means/debts.ts]: not much gain, a lot of pain ;)

File: src/main/features/response/views/statement-of-means/debts.njk (hunk 5 above)
Comment (yarpo) [reply src/main/features/response/views/statement-of-means/debts.njk]: the thing that we reuse everywhere :/ Should be renamed to more generic thing. 

I hope I will have time to retrospectively refactor old multi-row forms (such us claim amount, evidence and timeline). I want to change this css then as well.

File: src/test/app/forms/validation/validators/atLeastOnePopulatedRow.ts [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+
+describe('AtLeastOnePopulatedRowConstraint', () => {
+  const constraint: AtLeastOnePopulatedRowConstraint = new AtLeastOnePopulatedRowConstraint()
+
+  describe('validate', () => {
+
+    context('should accept', () => {
+
+      it('should accept null value', () => {
+        expect(constraint.validate(null)).to.be.equal(true)
```
Comment (timja-kainos) [inline src/test/app/forms/validation/validators/atLeastOnePopulatedRow.ts]: null?

File: src/main/app/forms/validation/validators/atLeastOnePopulatedRow.ts [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+} from 'class-validator'
+
+import { MultiRowForm } from 'forms/models/multiRowForm'
+import { MultiRowFormItem } from 'forms/models/multiRowFormItem'
+
+@ValidatorConstraint()
+export class AtLeastOnePopulatedRowConstraint implements ValidatorConstraintInterface {
+
+  validate (value: any | MultiRowForm<MultiRowFormItem>, args?: ValidationArguments): boolean {
+    if (value == null) {
```
Comment (timja-kainos) [inline src/main/app/forms/validation/validators/atLeastOnePopulatedRow.ts]: null?

Comment (yarpo) [reply src/main/app/forms/validation/validators/atLeastOnePopulatedRow.ts]: we have line like this in all validators...

File: src/test/app/forms/validation/validators/atLeastOnePopulatedRow.ts (hunk 12 above)
Comment (yarpo) [reply src/test/app/forms/validation/validators/atLeastOnePopulatedRow.ts]: as below. We can try to change it in one PR in all validators at once.

File: src/main/app/forms/validation/validators/atLeastOnePopulatedRow.ts (hunk 13 above)
Comment (timja-kainos) [reply src/main/app/forms/validation/validators/atLeastOnePopulatedRow.ts]: Untrue, I checked most of them and approximately half (including the first 3 I checked from the top) all use undefined.
New code should follow the newer convention, and if you would like to update all of them afterwards great too

**Pull request merged**